code/datums/client_image_controller.dm

```diff
@@ -0,0 +1,359 @@
+		if(need_refresh && each_mob.client && !already_refreshed[each_mob])
+			_refresh_shared_client_images(each_mob) // recovering shared images from other image groups
+	already_refreshed.Cut()
+
+	// NOTE: please make sure if images are not referenced
+	cimg_holder.bound_images.Cut()
+	cimg_holder.shared_bound_images.Cut()
+
+	cimg_holder.valid_mobs.Cut()
+	cimg_holder.valid_minds.Cut()
+	cimg_holders -= cimg_key
+	qdel(cimg_holder)
+
+/// this actually refreshes every client images - aghost or getting a new mob will call this
+/datum/cimg_controller/proc/on_mob_log_on(mob/cimg_mob)
+	var/list/already_injected = list()
+	for(var/each_cimgkey in cimgkey_by_mob[cimg_mob])
+		var/datum/cimg_holder/cimg_holder = cimg_holders[each_cimgkey]
+		cimg_mob.client.images += cimg_holder.bound_images
+		cimg_mob.client.images |= cimg_holder.shared_bound_images
+		already_injected[each_cimgkey] = TRUE
+	if(!cimg_mob.mind)
+		return
+	for(var/each_cimgkey in cimgkey_by_mind[cimg_mob.mind])
+		if(already_injected[each_cimgkey])
+			continue
+		var/datum/cimg_holder/cimg_holder = cimg_holders[each_cimgkey]
+		cimg_mob.client.images += cimg_holder.bound_images
+		cimg_mob.client.images |= cimg_holder.shared_bound_images
+
+/// something was removed, and we check full shared images
+/datum/cimg_controller/proc/_refresh_shared_client_images(mob/cimg_mob)
+	var/list/already_injected = list()
+	for(var/each_cimgkey in cimgkey_by_mob[cimg_mob])
+		var/datum/cimg_holder/cimg_holder = cimg_holders[each_cimgkey]
+		cimg_mob.client.images |= cimg_holder.shared_bound_images
+		already_injected[each_cimgkey] = TRUE
+	if(!cimg_mob.mind)
+		return
+	for(var/each_cimgkey in cimgkey_by_mind[cimg_mob.mind])
+		if(already_injected[each_cimgkey])
+			continue
+		var/datum/cimg_holder/cimg_holder = cimg_holders[each_cimgkey]
+		cimg_mob.client.images |= cimg_holder.shared_bound_images
+
+/datum/cimg_controller/proc/on_mob_destroy(mob/cimg_mob)
+	for(var/each_cimgkey in cimgkey_by_mob[cimg_mob])
+		var/datum/cimg_holder/cimg_holder = cimg_holders[each_cimgkey]
+		cimg_holder.valid_mobs -= cimg_mob
+	cimgkey_by_mob -= cimg_mob
+
+/datum/cimg_controller/proc/on_mind_destroy(datum/mind/cimg_mind)
+	if(!istype(cimg_mind, /datum/mind))
+		CRASH("proc 'on_mind_destroy' has taken non-mind")
+	for(var/each_cimgkey in cimgkey_by_mind[cimg_mind])
+		var/datum/cimg_holder/cimg_holder = cimg_holders[each_cimgkey]
+		cimg_holder.valid_minds -= cimg_mind
+		// there's no way to track a mind's client.
+	cimgkey_by_mind -= cimg_mind
+
+/// Makes a mob can see images that are bound to a key group.
+/// NOTE: calling this again adds up +1 count the validation. call disqualify_mob() proc to handle this correctly.
+/datum/cimg_controller/proc/validate_mob(cimg_key, mob/cimg_mob)
+	if(!cimg_mob)
+		return
+	if(!ismob(cimg_mob))
+		if(istype(cimg_mob, /datum/mind))
+			CRASH("validate_mob() has taken a mind. You called a wrong proc.")
+		CRASH("validate_mob() has taken non-mob")
```

Comment:
Just say what the type of the thing being passed is and then you don't need explicit handling for /datum/mind